### PR TITLE
Use SecureTransport on OS X

### PR DESCRIPTION
**Pull request by carlosmn**
Apple provides a TLS communication library which is included with the OS, thead-safe by itself and generally a better sport than OpenSSL.

Custom certificate checking is still not implemented, but the other clone tests do work.

- [x] Get certificate extraction working
- [x] Better message in CMake about finding the frameworks
- [x] Rename `GIT_SSL` to `GIT_OPENSSL` as that's what it really means
- [x] Don't fail a handshake automatically on bad certificate

### CODE REVIEW

**Comment by carlosmn** [comment]
I'm pretty happy about not linking against OpenSSL at all.

```
% otool -L libgit2.dylib                                     
libgit2.dylib:
        /Users/carlos/git/libgit2/libgit2.22.dylib (compatibility version 22.0.0, current version 0.22.0)
        /usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
        /System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 55471.14.27)
        /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 855.17.0)
        /usr/local/lib/libssh2.1.dylib (compatibility version 2.0.0, current version 2.1.0)
        /usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
```

**Comment by carlosmn** [comment]
It looks like we stopped running the online tests on travis, since there's no way this can be green.

**Comment by Therzok** [comment]
So many :heart: for you.

**Comment by carlosmn** [comment]
There is still of course libssh2 which will still link against openssl, so we're not quite free of threading nonsense yet, but someone was trying to upstream a SecureTransport version.

**Comment by ethomson** [comment]
Man, I'm excited about having this.  Very clean implementation.

**Comment by carlosmn** [comment]
@Therzok @nulltoken btw you should be able to write an implementation of the `git_stream` interface in C# so you use a managed TLS stream and don't have to worry about what's available in the OS since the runtime would take care of it (part of the reason for writing this layer was to allow making the use of OpenSSL by libgit2 irrelevant by letting the bindings bypass it).

**Comment by Therzok** [comment]
@nulltoken Should we make the [magic](https://msdn.microsoft.com/en-us/library/system.net.security.sslstream%28v=vs.110%29.aspx) happen?

**Comment by Therzok** [comment]
@carlosmn I can see the [definition](https://github.com/libgit2/libgit2/blob/a36486ef26dac5f4f359cc54c865b91bbc31767e/include/git2/sys/stream.h#L28-L38), but can't find where to hook it. Any tips?

**Comment by carlosmn** [comment]
That function doesn't exist yet, as that depends on `git_tls_stream_new()` abstracting away the actual implementation first, which I only got around to in this PR.

My idea was to have a function like `git_tls_stream_set()` (or better) name which would set some global variable and then `git_tls_stream_new()` would look into that variable and call your constructor instead of the OpenSSL or SecureTransport one.

**Comment by Therzok** [comment]
Lgtm :+1:

git_tls_stream_set_factory()?

File: src/stransport_stream.c [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	if ((error = git_stream_connect(st->io)) < 0)
+		return error;
+
+	ret = SSLHandshake(st->ctx);
+	if (ret != errSSLServerAuthCompleted) {
+		giterr_set(GITERR_SSL, "unexpected return value from ssl handshake %d", ret);
+		return -1;
+	}
+
+	if ((ret = SSLCopyPeerTrust(st->ctx, &trust)) != noErr)
```
Comment (ethomson) [inline src/stransport_stream.c]: If this fails, is it possible for `trust` to remain unset (and then we subsequently try to `CFRelease` it at line 81)?  Which is to say should `trust` have a default?

**Comment by ethomson** [comment]
This looks great to me (aside from that one question).  VERY EXCITED FOR THIS.

File: src/stransport_stream.c (hunk 1 above)
Comment (carlosmn) [inline src/stransport_stream.c]: Yep, it should be `SecTrustRef trust = NULL`.

**Comment by ethomson** [comment]
e8nzc

**Comment by ethomson** [comment]
giphy

**Comment by ethomson** [comment]
animated

**Comment by ethomson** [comment]
We have successfully eliminated the scourge of OpenSSL on another platform.  All thanks be to @carlosmn .

**Pull request merged**